ui/src/asm_cleanup.rs

```diff
@@ -107,7 +110,16 @@ pub fn filter_asm(block: &str) -> String {
         }
     }
 
-    let used_labels: HashSet<_> = labels.intersection(&opcode_operands).collect();
+    let mut data_labels = Vec::new();
+    let mut used_labels: HashSet<_> = labels.intersection(&opcode_operands).collect();
+
+    // We only include labels ref'd by data dirs if directly or indirectly used by an opcode
+    used_labels.iter()
+        .filter(|label| label_graph.contains_node(label))
+        .flat_map(|n| label_graph.neighbors_directed(n, Outgoing))
+        .for_each(|node| data_labels.push(node));
```

Comment:
Very succinct! Could this now be

```rust
    let data_labels: Vec<_> = used_labels.iter()
        .filter(|label| label_graph.contains_node(label))
        .flat_map(|n| label_graph.neighbors_directed(n, Outgoing))
        .collect();
```

I noticed the loop is gone; do we no longer need to deeply trace the graph?

Reply:
Gah, you're right!  I apologize, let me take another pass at this.

Reply:
Thanks for putting up with all my pedantic quibbles ;-)

Reply:
I'm guessing the loop can be broken as:

```rust
if data_labels.is_empty() { 
    break;
}
```

Reply:
Actually, it's probably if `used_labels` doesn't change size when the `data_labels` are added...

Reply:
I don't consider "not working correctly" to be a pedantic quibble 😄 

It should be checking the full graph now, and I added a test, which I should have done earlier.